models/git/branch_list.go

```diff
@@ -0,0 +1,111 @@
+// Copyright 2023 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package git
+
+import (
+	"context"
+
+	"code.gitea.io/gitea/models/db"
+	user_model "code.gitea.io/gitea/models/user"
+	"code.gitea.io/gitea/modules/container"
+	"code.gitea.io/gitea/modules/util"
+
+	"xorm.io/builder"
+)
+
+type BranchList []*Branch
+
+func (branches BranchList) LoadDeletedBy(ctx context.Context) error {
+	ids := container.Set[int64]{}
+	for _, branch := range branches {
+		ids.Add(branch.DeletedByID)
+	}
+	usersMap := make(map[int64]*user_model.User, len(ids))
+	if err := db.GetEngine(ctx).In("id", ids.Values()).Find(&usersMap); err != nil {
+		return err
+	}
+	for _, branch := range branches {
+		branch.DeletedBy = usersMap[branch.DeletedByID]
+		if branch.DeletedBy == nil {
+			branch.DeletedBy = user_model.NewGhostUser()
+		}
+	}
+	return nil
+}
+
+// LoadAllBranches loads all branches of a repository from database includes deleted branches
+func LoadAllBranches(ctx context.Context, repoID int64) ([]*Branch, error) {
+	var branches []*Branch
+	err := db.GetEngine(ctx).Where("repo_id=?", repoID).
+		Find(&branches)
+	return branches, err
+}
+
+type FindBranchOptions struct {
+	db.ListOptions
+	RepoID               int64
+	IncludeDefaultBranch bool
+	IsDeletedBranch      util.OptionalBool
+}
+
+func (opts *FindBranchOptions) Cond() builder.Cond {
+	cond := builder.NewCond()
+	if opts.RepoID > 0 {
+		cond = cond.And(builder.Eq{"repo_id": opts.RepoID})
+	}
+	if !opts.IncludeDefaultBranch {
+		cond = cond.And(builder.Neq{"name": builder.Select("default_branch").From("repository").Where(builder.Eq{"id": opts.RepoID})})
```

Comment:
It seems that the caller should already know the "default branch"

Could the "cond" be something like "name NOT IN '[]ExcludedBranchNames'" ?